java/connector-node/risingwave-sink-cassandra/src/main/java/com/risingwave/connector/CassandraSink.java

```diff
@@ -0,0 +1,263 @@
+                        // cassandra does not allow SET on primary keys
+                        .filter(c -> !tableSchema.getPrimaryKeys().contains(c))
+                        .collect(Collectors.toList());
+
+        this.stmtMap = new HashMap<>();
+        // prepare statement for insert
+        this.stmtMap.put(
+                "insert", session.prepare(createInsertStatement(config.getTable(), tableSchema)));
+        if (config.getType().equals("upsert")) {
+            // prepare statement for update-insert/update-delete
+            this.stmtMap.put(
+                    "update",
+                    session.prepare(createUpdateStatement(config.getTable(), tableSchema)));
+
+            // prepare the delete statement
+            this.stmtMap.put(
+                    "delete",
+                    session.prepare(createDeleteStatement(config.getTable(), tableSchema)));
+        }
+    }
+
+    @Override
+    public void write(Iterator<SinkRow> rows) {
+        System.out.println(this.config.getType());
+        if (this.config.getType().equals("append-only")) {
+            write_append_only(rows);
+        } else {
+            write_upsert(rows);
+        }
+    }
+
+    private void write_append_only(Iterator<SinkRow> rows) {
+        while (rows.hasNext()) {
+            SinkRow row = rows.next();
+            Data.Op op = row.getOp();
+            switch (op) {
+                case INSERT:
+                    batchBuilder.addStatement(bindInsertStatement(this.stmtMap.get("insert"), row));
+                    break;
+                case UPDATE_DELETE:
+                    break;
+                case UPDATE_INSERT:
+                    break;
+                case DELETE:
+                    break;
+                default:
+                    throw Status.INTERNAL
+                            .withDescription("Unknown operation: " + op)
+                            .asRuntimeException();
+            }
+        }
+    }
+
+    private void write_upsert(Iterator<SinkRow> rows) {
+        while (rows.hasNext()) {
+            SinkRow row = rows.next();
+            Data.Op op = row.getOp();
+            switch (op) {
+                case INSERT:
+                    batchBuilder.addStatement(bindInsertStatement(this.stmtMap.get("insert"), row));
+                    break;
+                case UPDATE_DELETE:
+                    updateRowCache.clear();
+                    updateRowCache.add(row);
+                    break;
+                case UPDATE_INSERT:
+                    SinkRow old = updateRowCache.remove(0);
+                    if (old == null) {
+                        throw Status.FAILED_PRECONDITION
+                                .withDescription("UPDATE_INSERT without UPDATE_DELETE")
+                                .asRuntimeException();
+                    }
+                    batchBuilder.addStatement(
+                            bindUpdateInsertStatement(this.stmtMap.get("update"), old, row));
```

Comment:
FYI, cassandra is upsert by design. I think we can use insert statement instead of update statement here to simplify the codes. I am also okay of keeping the current implementation.